### PR TITLE
procfs-util: fix confusion wrt. quantity limit and maximum value

### LINKED ISSUE #21193: Flaky `test-process-util/test_get_process_ppid`
**systemd version the issue has been seen with**
 > latest master

**Used distribution**
 > Fedora Rawhide, Arch Linux, Ubuntu Focal

This issue seems to plague all CIs from what I've seen so far:

```
>>> MALLOC_PERTURB_=36 SYSTEMD_LANGUAGE_FALLBACK_MAP=/builddir/build/BUILD/systemd-249/src/locale/language-fallback-map SYSTEMD_KBD_MODEL_MAP=/builddir/build/BUILD/systemd-249/src/locale/kbd-model-map PATH=/builddir/build/BUILD/systemd-249
/redhat-linux-build:/builddir/.local/bin:/builddir/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/sbin /builddir/build/BUILD/systemd-249/redhat-linux-build/test-process-util
――――――――――――――――――――――――――――――――――――― ✀  ―――――――――――――――――――――――――――――――――――――
Listing only the last 100 lines from a long log.
stderr:
...
/* test_setpriority_closest */
/* test_get_process_ppid */
Assertion 'limit >= INT_MAX || get_process_ppid(limit+1, NULL) == -ESRCH' failed at src/test/test-process-util.c:855, function test_get_process_ppid(). Aborting.
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
```

### PR DESCRIPTION
**Pull request by keszybz**
From packit/rawhide-arm64 logs:
Assertion 'limit >= INT_MAX || get_process_ppid(limit+1, NULL) == -ESRCH' failed at src/test/test-process-util.c:855, function test_get_process_ppid(). Aborting.
â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•â€•

The kernel has a few different limits. In particular kernel.threads-max can be
set to some lower value, and kernel.pid_max can be set to a higher value. This
is nice because it reduces PID reuse, even if the number of threads that is
allowed is limited. But the tests assumed that we cannot have a thread with
PID above MIN(kernel.threads-max, kernel.pid_max-1), which is not valid.

So let's rework the whole thing: let's expose the helpers to read
kernel.threads-max and kernel.pid_max, and print what they return in tests.
procfs_tasks_get_limit() was something that is only used in tests, and wasn't
very well defined, so let's drop it.

Fixes #21193.

### CODE REVIEW

**Comment by bluca** [comment]
Computer says no:

```
20:23:52 #1  0x00007fc215a59db5 in abort () from /lib64/libc.so.6
20:23:52 No symbol table info available.
20:23:52 #2  0x00007fc215f70271 in log_assert_failed (
20:23:52     text=text@entry=0x4013e8 "r == 0 || ERRNO_IS_PRIVILEGE(r)", 
20:23:52     file=file@entry=0x40130b "src/test/test-procfs-util.c", 
20:23:52     line=line@entry=54, 
20:23:52     func=func@entry=0x401541 <__PRETTY_FUNCTION__.7242> "main")
20:23:52     at ../src/basic/log.c:866
20:23:52 No locals.
20:23:52 #3  0x0000000000401148 in main (argc=<optimized out>, argv=<optimized out>)
20:23:52     at ../src/test/test-procfs-util.c:54
20:23:52         nsec = 664610000000
20:23:52         v = 255266
20:23:52         w = 255266
20:23:52         r = -30
20:23:52         __PRETTY_FUNCTION__ = "main"
20:23:52         __func__ = "main"
```

File: src/basic/limits-util.c [hunk 1]
```diff
@@ -111,35 +111,57 @@ uint64_t physical_memory_scale(uint64_t v, uint64_t max) {
...
+        /* Determine the maximum number of tasks that may run on this system. We check three sources to
+         * determine this limit:
          *
-         * a) the maximum tasks value the kernel allows on this architecture
-         * b) the cgroups pids_max attribute for the system
-         * c) the kernel's configured maximum PID value
+         * a) kernel.threads-max sysctl: the maximum number of tasks (threads) the kernel allows.
          *
-         * And then pick the smallest of the three */
+         *    This puts a direct limit on on the number of concurrent tasks.
```
Comment (yuwata) [inline src/basic/limits-util.c]: "on on"

File: src/test/test-process-util.c [hunk 2]
```diff
@@ -851,8 +851,14 @@ static void test_get_process_ppid(void) {
...
 
         /* the process with the PID above the global limit definitely doesn't exist. Verify that */
-        assert_se(procfs_tasks_get_limit(&limit) >= 0);
-        assert_se(limit >= INT_MAX || get_process_ppid(limit+1, NULL) == -ESRCH);
+        assert_se(procfs_get_pid_max(&limit) >= 0);
+        log_debug("kernel.pid_max = %"PRIu64, limit);
+
+        if (limit < INT_MAX) {
+                r = get_process_ppid(limit + 1, NULL);
+                log_debug_errno(r, "get_process_limit(%"PRIu64") → %d/%m", limit + 1, r);
```
Comment (yuwata) [inline src/test/test-process-util.c:859]: Maybe, ":" is better for the separator?

File: src/test/test-procfs-util.c [hunk 3]
```diff
@@ -25,26 +26,39 @@ int main(int argc, char *argv[]) {
...
+        assert(r >= 0);
 
         if (v > 100) {
-                uint64_t w;
+                log_info("Reducing limit by one to %"PRIu64"…", v-1);
+
                 r = procfs_tasks_set_limit(v-1);
-                assert_se(IN_SET(r, 0, -EPERM, -EACCES, -EROFS));
+                log_info_errno(r, "procfs_tasks_set_limit: %m");
+                assert_se(r == 0 || ERRNO_IS_PRIVILEGE(r));
```
Comment (yuwata) [inline src/test/test-procfs-util.c]: `r >= 0` is better than `r == 0`.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/test/test-process-util.c (hunk 2 above)
Comment (keszybz) [reply src/test/test-process-util.c:859]: This is just debug output… I don't think it matters too much one way or the other.

**Comment by keszybz** [comment]
Thank you for the reviews, updated.

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/basic/limits-util.c [hunk 4]
```diff
@@ -111,35 +111,57 @@ uint64_t physical_memory_scale(uint64_t v, uint64_t max) {
...
+                r = cg_get_attribute_as_uint64("pids", root, "pids.max", &c);
                 if (r < 0)
-                        log_debug_errno(r, "Failed to read pids.max attribute of cgroup root, ignoring: %m");
+                        log_debug_errno(r, "Failed to read pids.max attribute of root cgroup, ignoring: %m");
         }
 
-        return MIN3(TASKS_MAX,
-                    a <= 0 ? TASKS_MAX : a,
-                    b <= 0 ? TASKS_MAX : b);
+        return MIN3(a, b, c);
```
Comment (poettering) [inline src/basic/limits-util.c:164]: I would have included TASKS_MAX in this expression too, even if all three of a/b/c could be determined...

**Comment by Werkov** [comment]
This seems to have broken tests on my machine (running with a4eba5d8cfaabbd87687c651fcdd06df9e267931). (Perhaps because because of running unprivileged?)

```
426/767 test-procfs-util                                                                   FAIL             2.24s   killed by signal 6 SIGABRT
>>> MALLOC_PERTURB_=178 SYSTEMD_LANGUAGE_FALLBACK_MAP=/root/src/src/locale/language-fallback-map PATH=/root/build:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin /root/build/test-procfs-util

stderr:
Current system CPU time: 2w 20h 32min 5.410000s
Current memory usage: 15.0G
Current number of tasks: 1388
kernel.pid_max: 65536
kernel.threads-max: 127297
Limit of tasks: 65535
Reducing limit by one to 65534…
procfs_tasks_set_limit: Read-only file system
Assertion 'r >= 0 ? w == v - 1 : w == v' failed at src/test/test-procfs-util.c:57, function main(). Aborting.
```

**Pull request merged**